### ISSUE TITLE
`Set Fast Name Printing.` breaks printing of notations that use `_`, such as `->`

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#10282]
#### Description of the problem

```
$ ~/.local64/coq/coq-master/bin/coqtop
Welcome to Coq jgross-Leopard-WS:/home/jgross/Documents/repos/coq,master (7032085c809993d6a173e51aec447c02828ae070)

Coq < Check Set -> Set.
Set -> Set
     : Type

Coq < Set Fast Name Printing.

Coq < Check Set -> Set.
forall H : Set, Set
     : Type

Coq <
```

Is this expected?

#### Coq Version

master (7032085c809993d6a173e51aec447c02828ae070)

cc @ppedrot?

**Comment by ppedrot** [comment]
I guess so. The fast name printing option doesn't take into account the actual use of bound names in subterms, so any notation relying on this won't work.

**Comment by JasonGross** [comment]
In that case, this caveat should be noted in the documentation.

I would be interested in an option that does occurrence checking to determine whether or not bound names are used at all (this should incur at most one additional walk over the term, as you can just maintain the set of currently bound variables and of variables seen so far, and have each recursive call return the set of used variables together with the properly anonymized term), assuming such an option still has a performance benefit; fiat-crypto currently globally sets `Fast Name Printing`, and the fact that it breaks `->` has me reconsidering this decision.

**Comment by ppedrot** [comment]
> this should incur at most one additional walk over the term, as you can just maintain the set of currently bound variables and of variables seen so far, and have each recursive call return the set of used variables together with the properly anonymized term

Nope, because this breaks the lazy detyping of terms. You have to eagerly force the term to return the set of used variables.
